torch_xla/csrc/xla_lower_util.cpp

```diff
@@ -1154,48 +1154,18 @@ xla::XlaOp BuildMultinomial(xla::XlaOp input, int64_t num_samples,
   one = BuildExpand(one, output_size);
   xla::XlaOp rng = RngUniform(seed, output_shape, zero, one);
 
-  // Map samples to categories: Start with highest category count and
-  // scan through the cumulative probabilities. decrement counter when
-  // sample is <= cumulative probability
-  // Note: Iterate backwards to prevent selecting a category with 0 probability.
-  auto output_type = xla::PrimitiveType::S64;
-  std::vector<int64_t> slice_sizes = XlaHelpers::SizesOfXlaOp(cumprob);
-  slice_sizes[dim] = 1;
-  xla::XlaOp max_index =
-      xla::ConstantR0<int64_t>(input.builder(), sizes[dim] - 1);
-  auto rank = slice_sizes.size();
-  xla::XlaOp output =
-      xla::ConstantR1<int64_t>(input.builder(), num_samples, sizes[dim]);
-  if (rank == 2) output = xla::BroadcastInDim(output, output_size, {1});
-
-  auto body_fn =
-      [&](xla::XlaOp counter, absl::Span<const xla::XlaOp> values,
-          xla::XlaBuilder* builder) -> xla::StatusOr<std::vector<xla::XlaOp>> {
-    auto output = values[0];
-    auto max_index = values[1];
-    auto cumprob = values[2];
-    auto rng = values[3];
-
-    xla::XlaOp prob_val;
-    if (rank == 2) {
-      xla::XlaOp zero = xla::ConstantR0<int64_t>(max_index.builder(), 0);
-      prob_val =
-          xla::DynamicSlice(cumprob, {zero, max_index - counter}, slice_sizes);
-    } else {  // rank == 1
-      prob_val = xla::DynamicSlice(cumprob, {max_index - counter}, slice_sizes);
-    }
-    xla::XlaOp rhs = XlaHelpers::ImplicitBroadcast(
-        prob_val, XlaHelpers::ShapeOfXlaOp(prob_val),
-        XlaHelpers::ShapeOfXlaOp(rng));
-    xla::XlaOp x = BuildComparisonOp(at::aten::le, rng, rhs);
-    output = output - xla::ConvertElementType(x, output_type);
-    return std::vector<xla::XlaOp>{output, max_index, cumprob, rng};
-  };
+  // Map samples to categories
+  std::vector<int64_t> broadcast_size = XlaHelpers::SizesOfXlaOp(input);
+  broadcast_size.push_back(num_samples);
+  cumprob = xla::BroadcastInDim(cumprob, broadcast_size, {dim});
+  rng = xla::BroadcastInDim(rng, broadcast_size, {input_shape.rank()});
 
-  std::vector<xla::XlaOp> results = ConsumeValue(xla::ForEachIndex(
-      sizes[dim], output_type, body_fn, {output, max_index, cumprob, rng},
-      "MultinomialMapping", output.builder()));
-  return results[0];
+  // Build comparison mask and sum along K dimension
+  auto output_type = xla::PrimitiveType::S64;
+  auto mask = BuildComparisonOp(at::aten::gt, rng, cumprob);
```

Comment:
This line seems to be responsible for the CI failure, related to shape differences between `rng` vs `cumprob`

Reply:
Yeah, the BroadcastInDim calls didn't have proper arguments for 2D inputs.